### PR TITLE
[optimize] Optimize spark load/broker load reading parquet format file

### PR DESCRIPTION
**Pull request by xy720**
Please see the description in #3877 .
spark load #3433

### CODE REVIEW

**Comment by yiguolei** [comment]
Great job

File: be/src/exec/buffered_reader.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <stdint.h>
+
+#include "common/status.h"
+#include "olap/olap_define.h"
+#include "exec/file_reader.h"
+
+namespace doris {
+
+// Buffered Reader of broker
```
Comment (kangkaisen) [inline be/src/exec/buffered_reader.h:28]: Please add the comment to describe why do we need this class, and when should we use it.

**Comment by kangkaisen** [comment]
@xy720 Hi, Please add the UT for new BufferedReader class, Thanks.

File: be/src/exec/buffered_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+BufferedReader::~BufferedReader() {
+    close();
+}
+
+Status BufferedReader::open() {
+    if (!_reader) {
+        std::stringstream ss;
+        ss << "Open buffered reader failed, reader is null";
+        LOG(WARNING) << ss.str();
```
Comment (morningman) [inline be/src/exec/buffered_reader.cpp:55]: Let the caller print the log, not here.

File: be/src/exec/buffered_reader.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+class BufferedReader : public FileReader {
+public:
+    // If the reader need the file size, set it when construct BrokerReader.
```
Comment (morningman) [inline be/src/exec/buffered_reader.h:31]:
```suggestion
    // If the reader need the file size, set it when construct FileReader.
```

File: be/src/exec/buffered_reader.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    virtual Status readat(int64_t position, int64_t nbytes, int64_t* bytes_read, void* out) override;
+    virtual Status read_one_message(uint8_t** buf, size_t* length) override;
+    virtual int64_t size() override;
+    virtual Status seek(int64_t position) override;
+    virtual Status tell(int64_t* position) override;
+    virtual void close() override;
+    virtual bool closed() override;
+
+private:
+    Status fill();
```
Comment (morningman) [inline be/src/exec/buffered_reader.h:50]:
```suggestion
    Status _fill();
```

File: be/src/exec/buffered_reader.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    Status read_once(int64_t position, int64_t nbytes, int64_t* bytes_read, void* out);
```
Comment (morningman) [inline be/src/exec/buffered_reader.h:51]:
```suggestion
    Status _read_once(int64_t position, int64_t nbytes, int64_t* bytes_read, void* out);
```

File: be/src/exec/buffered_reader.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return Status::InternalError(ss.str());
+    }
+    RETURN_IF_ERROR(_reader->open());
+    RETURN_IF_ERROR(fill());
```
Comment (morningman) [inline be/src/exec/buffered_reader.cpp:59]: Is it necessary to fill() when open()?

File: be/src/exec/buffered_reader.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    memcpy(out, _buffer + off, len);
+    *bytes_read = len;
+    _cur_offset = position + *bytes_read;
+    return Status::OK();
+}
+
+Status BufferedReader::fill() {
+    if (_buffer_offset >= 0) {
+        int64_t bytes_read;
+        int retry_times = 1;
```
Comment (morningman) [inline be/src/exec/buffered_reader.cpp:128]: What is this retry for? add comment for it.

File: be/src/exec/buffered_reader.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    // requested bytes missed the local buffer
+    if (position >= _buffer_limit || position < _buffer_offset) {
+        // if requested length is larger than the capacity of buffer, do not
+        // need to copy the character into local buffer.
+        if (nbytes > _buffer_limit - _buffer_offset) {
+            return _reader->readat(position, nbytes, bytes_read, out);
+        }
+        _buffer_offset = position;
+        RETURN_IF_ERROR(fill());
+        if (position >= _buffer_limit) {
```
Comment (morningman) [inline be/src/exec/buffered_reader.cpp:112]: when will `position > _buffer_limit`?

File: be/src/exec/buffered_reader.cpp (hunk 6 above)
Comment (xy720) [reply be/src/exec/buffered_reader.cpp:59]: Yes. If not call fill() in open(), the actual length of the buffer will be 0 when user call read().

File: be/src/exec/buffered_reader.cpp (hunk 7 above)
Comment (xy720) [reply be/src/exec/buffered_reader.cpp:128]: ok

File: be/src/exec/buffered_reader.h (hunk 5 above)
Comment (xy720) [reply be/src/exec/buffered_reader.h:51]: done

File: be/src/exec/buffered_reader.h (hunk 4 above)
Comment (xy720) [reply be/src/exec/buffered_reader.h:50]: done

File: be/src/exec/buffered_reader.h (hunk 3 above)
Comment (xy720) [reply be/src/exec/buffered_reader.h:31]: done

File: be/src/exec/buffered_reader.h (hunk 1 above)
Comment (xy720) [reply be/src/exec/buffered_reader.h]: ok

File: be/test/exec/buffered_reader_test.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    MonotonicStopWatch watch;
+    watch.start();
+    bool eof = false;
+    size_t read_length = 0;
+    while (!eof) {
+        size_t buf_len = 32 * 1024;
+        st = reader.read(buf, &buf_len, &eof);
+        ASSERT_TRUE(st.ok());
+        read_length += buf_len;
+    }
```
Comment (morningman) [inline be/test/exec/buffered_reader_test.cpp:52]: You need to check the result, or the unit test is meaningless

File: be/test/exec/buffered_reader_test.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    BufferedReader reader(&file_reader, 128 * 1024);
+    auto st = reader.open();
+    ASSERT_TRUE(st.ok());
+    uint8_t buf[10];
+    bool eof = false;
+    size_t buf_len = 10;
+
+    st = reader.read(buf, &buf_len, &eof);
+    ASSERT_TRUE(st.ok());
+    ASSERT_STREQ("bdfhjlnprt", std::string((char*)buf, buf_len).c_str());
```
Comment (morningman) [inline be/test/exec/buffered_reader_test.cpp:69]: check the `buf_len` after `read()`

File: be/src/exec/buffered_reader.cpp [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    return Status::OK();
+}
+
+Status BufferedReader::_read_once(int64_t position, int64_t nbytes, int64_t* bytes_read, void* out) {
+    // requested bytes missed the local buffer
+    if (position >= _buffer_limit || position < _buffer_offset) {
+        // if requested length is larger than the capacity of buffer, do not
+        // need to copy the character into local buffer.
+        if (nbytes > _buffer_limit - _buffer_offset) {
```
Comment (morningman) [inline be/src/exec/buffered_reader.cpp:106]: should it be `nbytes > _buffer_size ` ?

Comment (xy720) [reply be/src/exec/buffered_reader.cpp:106]: You are right.

**Review by morningman — DISMISSED** [review]
LGTM

File: be/src/exec/broker_reader.cpp [hunk 12]
```diff
@@ -155,6 +155,8 @@ Status BrokerReader::readat(int64_t position, int64_t nbytes, int64_t* bytes_rea
             return status;
         }
 
+        LOG(DEBUG) << "send readat request to broker:" << broker_addr << " position:" << position << ", read bytes length:" << nbytes;
```
Comment (wutiangan) [inline be/src/exec/broker_reader.cpp:158]: what is "readat ?"

Comment (morningman) [reply be/src/exec/broker_reader.cpp:158]: > what is "readat ?"

Like `pread`

Comment (xy720) [reply be/src/exec/broker_reader.cpp:158]: Yes, it's confusing.  Better named it as pread request.

File: be/src/exec/buffered_reader.cpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+// buffered reader
+
+BufferedReader::BufferedReader(FileReader* reader)
+        : _reader(reader),
+          _buffer_size(1024 * 1024),
+          _buffer_offset(0),
+          _buffer_limit(0),
+          _cur_offset(0) {
+    _buffer = new char[_buffer_size];
+}
```
Comment (wutiangan) [inline be/src/exec/buffered_reader.cpp:36]: maybe this construct funciton can be remove, for this is duplicated code.
you can add default paremerts to constructor function bellow.

Comment (xy720) [reply be/src/exec/buffered_reader.cpp:36]: good

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**